### PR TITLE
Stratimikos: New Tpetra stratimikos ex/test

### PR DESCRIPTION
**Pull request by jennloe**
@trilinos/stratimikos 

## Motivation
It came to my attention that Tpetra + Stratimikos + Belos were not tested anywhere without the inclusion of other packages (e.g. Galeri, MueLu). I have added new Tpetra/Stratimikos/Belos tests to complete this testing coverage. One test mimics prior Epetra testing; the other uses an arbitrary input list. This will be useful in a) future Epetra removal and b) addition of new mixed-precision capabilities.

Furthermore, I cleaned up the Stratimikos/test directory, removing some unused files.  I also renamed files to make it very clear which linear algebra backend each test is using.

### CODE REVIEW

File: packages/stratimikos/test/test_tpetra_belos_ex.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+//
+// ***********************************************************************
+// @HEADER
+
+#include "Teuchos_ParameterList.hpp"
+#include "Teuchos_VerboseObject.hpp"
+#include "Teuchos_CommandLineProcessor.hpp"
+#include "Teuchos_XMLParameterListHelpers.hpp"
+
+#include "Tpetra_Core.hpp"
```
Comment (cgcgcg) [inline packages/stratimikos/test/test_tpetra_belos_ex.cpp:47]: Do we need to declare a package dependency on Tpetra to use these headers?

Comment (bartlettroscoe) [reply packages/stratimikos/test/test_tpetra_belos_ex.cpp:47]: @cgcgcg, it has a dependency implicitly through the package `ThyraTpetraAdapters`:

https://github.com/trilinos/Trilinos/blob/5301398ff4e35da9c1fb92e0cf83e477e27b28fe/packages/stratimikos/cmake/Dependencies.cmake#L4-L8

It seems impossible for the package `ThyraTpetraAdapters` to not be pulling in Tpetra headers in its header files.  But, in general, yes there should be a listed dependency on `TpetraCore`.

NOTE: A future version of TriBITS will support public and private package dependencies.  (Currently, all package dependencies in TriBITS are public.)

**Comment by bartlettroscoe** [comment]
@jennloe, thanks for putting in the PR!  I was not at SNL when people started using Tpetra with Stratimikos but I have been glad to see that the interfaces worked without modification.

I will put in my full review after we get a around of PR testing to run and complete.

**Comment by bartlettroscoe** [comment]
Sorry for the delay, I will get my review in later today.

**Comment by jennloe** [comment]
> Sorry for the delay, I will get my review in later today.

Thanks!  I should be able to work on it this afternoon.

**Comment by bartlettroscoe** [comment]
I will post my review next but I just wanted to log the impacted tests were run and passed (except for some Galeri tests requiring `float` or `complex<float>`).

The following existing tests were renamed:

* `test_single_{+epetra+}_stratimikos_solver_driver_amesos` (2 tests)
* `test_single_{+epetra+}_stratimikos_solver_driver_aztecoo_ifpack` (2 tests)
* `test_single_{+epetra+}_stratimikos_solver_driver_aztecoo_ml`
* `test_single_{+epetra+}_stratimikos_solver_driver_aztecoo_np`
* `test_single_{+epetra+}_stratimikos_solver_driver_belos_ifpack`
* `test_single_{+epetra+}_stratimikos_solver_driver_belos_ml`
* `test_single_{+epetra+}_stratimikos_solver_driver_belos_np`

The following existing `Galeri` tests:

* `Galeri_double_Jacobi`
* `Galeri_double_Jacobi_Tpetra`
* `Galeri_double_Jacobi_HalfPrecision`
* `Galeri_float_Jacobi`
* `Galeri_complex_double_Jacob`
* `Galeri_complex_float_Jacobi`

had their executable driver renamed from `GaleriDriver` to `GaleriXpetraDriver`.  (But many of those require enable of complex and/or float which are not enabled in PR testing.)

Also, the new tests added where:

* `test_tpetra_stratimikos_solver_driver`
* `test_tpetra_belos_ex`
* `test_tpetra_belos_sing_reduce`

That is a total of 18 tests in a single build impacted by this PR.

We can see evidence of 15 of these tests being run and passing on CDash for the `cuda-11.4.2` build in [this query](https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&filtercount=6&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=buildname&compare2=65&value2=PR-11094-test&field3=buildname&compare3=64&value3=no-complex&field4=buildstarttime&compare4=83&value4=2022-10-10T18%3A03%3A08%20UTC&field5=buildstarttime&compare5=84&value5=2022-10-10T18%3A07%3A47%20UTC&field6=block&field6count=3&field6field1=testname&field6compare1=65&field6value1=Stratimikos_test_single_epetra_stratimikos_solver&field6field2=testname&field6compare2=65&field6value2=Stratimikos_test_tpetra&field6field3=testname&field6compare3=65&field6value3=Stratimikos_Galeri) which shows:

image

We don't see all of the `Stratimikos_Galeri_xxx` tests because the following 3 tests require `float` and/or `complex<float>` to be enabled:

* `Galeri_double_Jacobi_HalfPrecision`
* `Galeri_complex_float_Jacobi`
* `Galeri_float_Jacobi`

and those are not enabled in **any** of the Trilinos PR builds.

So that verifies that all of the impacted tests all ran and passed (except for the tests that never run in Trilinos PR testing).

**Review by bartlettroscoe — APPROVED** [review]
Thanks for putting this PR together!  Stratimikos really needed native tests with Tpetra.

I reviewed each commit individually and applied tailored git diffs to highlight what changed more clearly.  (This made it easier to review every single line that was changed or added).  And with that analysis, the changes and additions are straightforward.

I really like how the change-sets are broken into different commits according to [Separate Changes](https://linux.die.net/man/7/gitworkflows). Thanks for doing that!

This PR is okay to merge but I have a few comments/suggestions that may improve the git commits and the software:

**A)** The `epetra` and `tpetra` file names are not quite consistent, yet look to be about identical (i.e. a copy and paste with some edits).  For example, it looks like the files `test_single_epetra_stratimikos_solver_driver.cpp` and `test_tpetra_stratimikos_solver_driver.cpp` are identical except for:

```
$ diff packages/stratimikos/test/test_single_epetra_stratimikos_solver_driver.cpp packages/stratimikos/test/test_tpetra_stratimikos_solver_driver.cpp 
42c42,43
< #include "test_single_epetra_stratimikos_solver.hpp"
---
> #include <Tpetra_Core.hpp>
> #include "test_tpetra_stratimikos_solver.hpp"
52,53d52
<   Teuchos::GlobalMPISession mpiSession(&argc,&argv);
< 
58a58
> 
62a63,64
>   Tpetra::ScopeGuard tpetraScope(&argc,&argv);
>   {
77c79
<       "Testing program for Trilinos (and non-Trilinos) linear solvers access through Thyra."
---
>       "Testing program for Tpetra-Belos linear solvers access through Thyra."
92c94
<       = Thyra::test_single_epetra_stratimikos_solver(
---
>       = Thyra::test_tpetra_stratimikos_solver(
95a98
>   } //End Tpetra scope
```

Also, the file `test_single_tpetra_stratimikos_solver.cpp` in 165 lines long but only differs in 14 lines from the file `test_tpetra_stratimikos_solver.cpp` as shown by:

```
$ cd Trilinos/packages/stratimikos/

$ cat test/test_tpetra_stratimikos_solver.cpp | wc -l
165

$ diff test/test_single_epetra_stratimikos_solver.cpp test/test_tpetra_stratimikos_solver.cpp  | grep "^>" | wc -l
14
```

and if you view these files with `vim -d <file1> <file2>` you can really see they are copy and paste with just a few lines changed for the transition from Epetra to Tpetra.

Therefore, it seems like either `single_` should be taken out of `epetra` file names or `single_` should be added to the `tpetra` file names for consistency.  And this would make it more clear the relationship between these sets of files and allow direct comparisons between Epetra and Tpetra.

**B)** In general, one would squash commits like https://github.com/trilinos/Trilinos/pull/11094/commits/005ff79484e6495ce69dc3593a3c859ac9a69f8f and https://github.com/trilinos/Trilinos/pull/11094/commits/367570faff4a8cd1b0f2dfae53ec5311e550599e against their upstream commits where the typos were created.  That just takes a couple of minutes using `git rebase -i <base-sha1>` and it makes the commits free-standing and passing tests and makes reviews easier.

**C)** The file `stratimikos/test/test_tpetra_belos_ex.cpp` looks to be more of an example and not a test as it looks very similar to the file `stratimikos/example/simple_stratimikos_example.cpp` (like a copy and paste).  Should the file `test_tpetra_belos_ex.cpp` and the logic in the `CMakeLists.txt` file to build and run this executable not be moved to the `stratimikos/example/` directory for consistency?

Again, sorry for the delay in the review.  (I don't always practice what I preach :-( )

**Comment by jennloe** [comment]
Thanks, Ross!
Replies to the main comments:

A) Sure, I can make the naming consistent.  Wasn't too sure why "single" was in the name to begin with, so I'll remove it in my follow-on PR. 
B) Good point for next time.
C) As is, I think there is a fair case for moving this to the example file.  HOWEVER, my next PR will add some new features, and I need this version (not the other Tpetra driver) to test my features.  So it will be a true test shortly.

File: packages/stratimikos/test/CMakeLists.txt [hunk 2]
```diff
@@ -169,30 +166,89 @@ IF (${PROJECT_NAME}_ENABLE_Tpetra)
...
+    galeri_xpetra_driver.cpp
     COMM serial mpi
     )
 
   IF (Tpetra_INST_DOUBLE)
 
     TRIBITS_ADD_TEST(
-      GaleriDriver
+      GaleriXpetraDriver
       NAME "Galeri_double_Jacobi"
```
[the remark this answers is not in the document]
Comment (jennloe) [reply packages/stratimikos/test/CMakeLists.txt:245]: Sure.  I can fix this in the next PR.

File: packages/stratimikos/test/CMakeLists.txt [hunk 3]
```diff
@@ -252,9 +308,10 @@ ENDIF ()
 
 IF (${PACKAGE_NAME}_ENABLE_Ifpack2 AND ${PACKAGE_NAME}_ENABLE_ThyraTpetraAdapters AND Tpetra_INST_DOUBLE AND Tpetra_INST_INT_INT)
  TRIBITS_ADD_EXECUTABLE(
-   issue_535
+   tpetra_ifpack2_issue_535
```
[the remark this answers is not in the document]
Comment (jennloe) [reply packages/stratimikos/test/CMakeLists.txt:311]: Looks like it should have been a test?  Should I make it a test?

**Comment by bartlettroscoe** [comment]
> B) Good point for next time.

@jennloe, if you are going to add new commits, it will fire off a new round of PR builds so you might as well run `git rebase -i` to clean up the commits.  If you are not sure how to do that, I can do that for you to demonstrate.

File: packages/stratimikos/test/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (bartlettroscoe) [reply packages/stratimikos/test/CMakeLists.txt:311]: Might be worth a try.  But if it fails, you may have to disable it (with the DISABLE test property to make it clear that the test should be running but is now and why).

**Comment by jennloe** [comment]
@bartlettroscoe I'll be touching this directory again in my next PR. Your comments are minor enough that I will fix those along with the new features I am adding. Thanks!

**Pull request merged**